### ISSUE TITLE
Add support for arbitrary objects defining a __mpz__ method

### ISSUE THREAD

**Issue by jdemeyer** [aleaxit/gmpy#123]
It should be possible for other projects to define a conversion to `gmpy2` types. As is typically done in Python, this could be done by a `__mpz__` method. That is, the function `mpz` should do something like (assuming also #124):

``` python
def mpz(x):
    # Existing code
    if isinstance(x, known_types):
        return known_conversion_to_mpz(x)

    # New code
    try:
        m = x.__mpz__
    except AttributeError:
        pass
    else:
        return m()
    try:
        x = int(x)
    except TypeError:
        pass
    else:
        return known_conversion_to_mpz(x)

    # Existing code
    raise TypeError("mpz() requires numeric or string argument")
```

**Comment by videlec** [comment]
See #134

**Comment by videlec** [comment]
this could be closed now that #134 is merged...

**Comment by videlec** [comment]
this could be closed...

**Comment by casevh** [comment]
Closing. Changes merged in #134.
